Check paths for privacy as they are resolved and refactor away `LastPrivate`

jseyfried opened the pull request:
This PR privacy checks paths as they are resolved instead of in `librustc_privacy` (fixes #12334 and fixes #31779). This removes the need for the `LastPrivate` system introduced in PR #9735, the limitations of which cause #31779.

This PR also reports privacy violations in paths to intra- and inter-crate items the same way -- it always reports the first inaccessible segment of the path.

Since it fixes #31779, this is a [breaking-change]. For example, the following code would break:
```rust
mod foo {
    pub use foo::bar::S;
    mod bar { // `bar` should be private to `foo`
        pub struct S;
    }
}

impl foo::S {
    fn f() {}
}

fn main() {
    foo::bar::S::f(); // This is now a privacy error
}
```

r? @alexcrichton

jseyfried commented:
cc @nrc @petrochenkov @nikomatsakis

petrochenkov commented:
Reviewed.

alexcrichton commented:
Thanks for the PR @jseyfried! I may unfortunately not be the best reviewer for this as so much of resolve and privacy have changed since this was originally implemented, but I can at least add my 2cents!

Historically privacy and resolve were entangled, but this ended up leading to tons of really weird resolution problems. The major motivation for moving all of privacy to its own separate pass was to improve all these errors. At that point, the only privacy-related parts of resolve were that glob imports didn't bring private names into a module. 

Since then, though, maybe resolve has changed to be much more robust? I've definitely seen some awesome work of yours fly by in the interim :).

I suspect that the buggy program you've linked here can possibly be fixed by not jettisoning `LastPrivate` but patching up support for it. It looks like this approach may be cleaner, though? It at least has a negative diffstat!

alexcrichton commented:
Looking over the tests, it looks like the error messages related to privacy get *even better* than where they are today. Specifically the messages seem much more precise in terms of what's actually private along the chain of a path.

With that in mind, if you're confident that this fits nicely into resolve as-is today and it won't regress resolution error messages, then it seems like a fine shift to me! I forget now what the `rustc_privacy` pass would actually do if this is removed, but it seems large enough that it surely does *something*.

jseyfried commented:
@alexcrichton Thanks for the history and feedback! Resolve is indeed more robust now :)

I'm confident that this PR won't regress resolution error messages since it doesn't change the observable behavior of the import resolver or crate resolver; instead, it silently collects privacy errors during resolution and reports them after resolution is complete.

`rustc_privacy` still does a couple of passes that are unaffected by this PR, including checking for public types in private interfaces. The pass that used to privacy check paths, `PrivacyChecker`, still privacy checks struct fields (including the sometimes implicit fields in struct constructors) and non-UFCS methods. Field checking can be moved into `resolve` and non-UFCS method checking can be moved into `typeck`, but that is for another PR.

jseyfried commented:
> I may unfortunately not be the best reviewer for this

r? @nikomatsakis

jseyfried commented:
I added a commit addressing @petrochenkov's comments.
I tested the commit locally -- travis is isn't working right now.

gereeter commented on src/librustc_resolve/lib.rs:
```diff
@@ -3614,6 +3606,12 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
     }
+
+    fn report_privacy_errors(&self) {
+        for (span, string) in self.privacy_errors.iter() {
```
Since `privacy_errors` is a `HashMap`, this will result in nondeterministic error ordering. Maybe the map should be changed to an `FnvHashMap` or a `BTreeMap`?

jseyfried replied:
Good point, I'll change it to a `BTreeMap`.

bors commented:
:umbrella: The latest upstream changes (presumably #31882) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
I've added a couple more comments since you are going to rebase this anyway.

nikomatsakis commented:
This seems like a big improvement. r=me.

nikomatsakis commented:
To be clear: r=me after rebase + nits that others have already listed are addressed.

jseyfried commented:
@nikomatsakis I rebased, addressed the comments, and cleaned up the history. Given the breakage in `rustc_metadata`, do you think this needs a crater run?

nikomatsakis commented:
@jseyfried unclear, but I kicked off one just in case

nikomatsakis commented:
The crater run somehow failed. I'm inclined to just land this, since the next beta has branched.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit e67590b has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis e67590b0821c8985a02b699471d85be6d2137bff -->

bors commented:
:hourglass: Testing commit e67590b with merge 7cee8b9...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8268), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8291), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1539), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8231), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8238), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8231), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1438), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3392), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8277), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8279), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8247), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/407), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3287), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3303), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3295), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3277), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2509), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3281)
